recipes-extended/xen/files/libxl-stubdom-allow-msi-enable.patch

```diff
@@ -0,0 +1,20 @@
+--- a/tools/libxl/libxl_pci.c
++++ b/tools/libxl/libxl_pci.c
+@@ -1417,6 +1417,17 @@ static void pci_add_dm_done(libxl__egc *
+             rc = ERROR_FAIL;
+             goto out;
+         }
++    } else if (libxl_is_stubdom(ctx, domid, NULL)) {
```

Comment:
This is in support of the HVM ndvm usecase?

Reply:
In the 256b62a3224423f6293163dde7e1d6f7eb3801fd commit message, I wrote
    Add this Qubes patch to allow stubdoms to enable MSI.  Without it,
    e1000e were reporting that they could not enable MSI and falling back to
    regular IRQs.

Qubes upstreamed some other changes to how IRQs are handled.  I think this piece hasn't gone in since it's for linux stubdom pci passthrough which isn't upstream.

Reply:
d7bcc139d6d61c750f234db278ad9a28bcfadda0 on this branch.

Reply:
We don't see e1000e complaining today though, even in the HVM ndvm case.  Looking at a recent openxt build
```
00:04.0 Ethernet controller: Intel Corporation Ethernet Connection (11) I219-LM
	Subsystem: Hewlett-Packard Company Device 8753
	Flags: bus master, fast devsel, latency 64, IRQ 56
	Memory at f1000000 (32-bit, non-prefetchable) [size=128K]
	Capabilities: [c8] Power Management version 3
	Capabilities: [d0] MSI: Enable+ Count=1/1 Maskable- 64bit+
	Kernel driver in use: e1000e
	Kernel modules: e1000e
```
MSI: Enable+, and nothing in dmesg concerning MSI.

Reply:
I may have been running with pci_permissive=0 - I've been testing that off and on.  With pci_permissive=1, the guest can enable MSI-X in the PCI config space, but that is denied with pci_permissive=0.  This is the Qubes commit adding the patch: https://github.com/QubesOS/qubes-vmm-xen/commit/9774d93bb602052c1eb16f1bc9639e3edb16ed06

Supporting pci_permissive=0 is something I'd like to introduce as at least an option.  Should I drop this for the time being?  For xenmgr, my quick patch just makes a per-VM pci_permissive setting.  Per-PCI device seems like it would be better, but implementing that is more complicated.

Reply:
It's not directly needed to support 4.16, and I also think it deserves having its own PR(s).